### PR TITLE
feat!: Infer input from source, and allow typing directly

### PR DESCRIPTION
**Pull request by jquense**
BREAKING CHANGE: types are incompatible

### CODE REVIEW

**Comment by jquense** [comment]
I figured now was the moment for that anyway, tbh most of usages are already broken

File: src/types/createResolve.ts [hunk 1]
```diff
@@ -2,13 +2,13 @@ import { GraphQLFieldConfig, GraphQLFieldResolver } from 'graphql';
 
 import NodeType from './NodeType';
 
-export default function createResolve<TFields extends string, TContext = any>(
-  resolve: GraphQLFieldResolver<Record<TFields, any>, TContext>,
-  fieldNames: TFields[],
-): GraphQLFieldConfig<any, TContext>['resolve'] {
+export default function createResolve<TSource, TContext = any>(
```
Comment (ethanve) [inline src/types/createResolve.ts:5]: Maybe the only thing we should do is `TSource extends {id: string;}`?

File: src/types/createResolve.ts [hunk 2]
```diff
@@ -2,13 +2,13 @@ import { GraphQLFieldConfig, GraphQLFieldResolver } from 'graphql';
...
+  resolve: GraphQLFieldResolver<TSource, TContext>,
+  ensureFields: (keyof TSource)[],
+): GraphQLFieldConfig<TSource, TContext>['resolve'] {
```
Comment (ethanve) [inline src/types/createResolve.ts:8]: One more thing -- won't this make the resolve's object argument be typed with all the source fields? 
Ideally we want the object to be `keys of ensureFields` and `id`

**Comment by jquense** [comment]
In practice TSource is not typed as a subset, e.g. `NodeType.getNodeObject` returns `TSource`, and everything at _least_ assumes TSource is a subset of TResolved which isn't strictly true (tho mostly is). I do see that it should be not infer from TSource tho, you'd only use this helper if it _wasn't_ in TSource

I really dislike how fuzzy 'obj' is in general in GraphQL, but that's a different thing...

Honestly tho I don't really understand the complexity of this helper, It's a lot of user specification for what is essentially just "Fetch this object if it's not already fetched", how often is it the case that that some stub contains more than 'id' where we need to individually specify which fields should exist? Is it not enough to cache node loads?

File: src/types/createResolve.ts (hunk 2 above)
Comment (jquense) [reply src/types/createResolve.ts:8]: the object that is passed in isn't jsut the fields you requested it's always the full node. The fields array is a maybe-to-complex hint to the helper on whether it needs to fetch the node or not. e.g. if `favoriteColor` is already on the source node don't make a request

**Comment by itajaja** [comment]
> Honestly tho I don't really understand the complexity of this helper, It's a lot of user specification for what is essentially just "Fetch this object if it's not already fetched".

yeah agree, but that wouldn't necessarily be easier to specify, unless you were going to give `any` to the obj, in which case you are losing extra checking. otherwise you'd still be forced to specify the field. so either way not great. this is really the main conceptual fallacy of the whole gnr thing

**Comment by jquense** [comment]
It seems like for Nodes, we should default to typing TSource as mostly `unknown`? Or actually i guess the important thing is it be the input to `makeId` otherwise there are really two generics that needed here `TInput` and `TNode` which would be returned from Resource. I get though that we don't want to explicitly and fully type the endpoint returns...we should generate typescript from Marshmallow schema or something :P

Overall tho @taion is is necessary to check if sources have fields? It seems like letting dataLoader handle the request optimization and making the helper more like "fetchThenResolve()` would be conceptually simpler?

**Comment by itajaja** [comment]
`fetchThenResolve` might fetch something that already exist. and unless you type the fetched obj, then it'd be any

**Comment by jquense** [comment]
> fetchThenResolve might fetch something that already exist.

what might exist? if it exists on the Source, don't use `fetchThenResolve` right? Generally tho how often is it the case that an input is **either** an object stub or a node? It seems like if a parent resolves something to a stub you always need to fetch the child node, it's not going to sometimes return the full object right?

**Comment by taion** [comment]
this was an optimization for two cases.

suppose you have a fragment like

```graphql
fragment WidgetFragment on Widget {
  name
  tenant {
    id
  }
}
```

so in this case we can resolve this without actually fetching the tenant. here's the other use case:

```graphql
fragment WidgetFragment on Widget {
  name
  tenant {
    id
    factories {
      name
    }
  }
}
```

this lets me fetch e.g. `/factories/?tenant_id=<tenant_id>` before waiting to resolve `/tenants/<tenant_id>`.

now, it turns out that, most likely, neither of these are that important.

that said, we have a distinct gap in our caching here of responses – items retrieved from "list" operations don't end up in the by-id cache, and neither do things like "get tenant by slug", because we don't really have a first-class concept of "this is an operation that retrieves an object of the type of the node resource".

my thought had been that i'd build a new library to handle this, but that was because i thought it was infeasible to migrate this library to typescript, and i didn't want to write more flow.

that said, now that we are on TS, i'd maybe lean toward building out some different API that merges the resources and the nodes, and maybe using that as the thing to fix caching, as opposed to necessarily changing/breaking the semantics on the existing constructs.

but i might be wrong there. it'd be nice to unify object types and resources, though.

**Pull request closed without merging**